airflow/opt/providers/etna/etna/hooks/c4.py

```diff
@@ -95,48 +98,47 @@ def __init__(self, hook: C4Hook):
 
     def upload_file(
         self,
+        sftp: SFTPClient,
         folder_path: str,
         file_name: str,
-        file_obj: io.BytesIO,
-        file_size: int
+        source_file: SftpEntry,
+        cat_hook: CatHook
      ):
         """
         Stream the file to C4, to the given path.
 
         params:
+          sftp: a connected SFTPClient instance
           folder_path: str, the destination path on C4 (relative to the hook's configured root_path)
           file_name: str, the file name to save to, on C4
-          file_obj: io, the file-like object that is a data source
-          file_size: int, size of the file in bytes
+          source_file: SftpEntry, the source file to download
+          cat_hook: CatHook, the CAT hook with SFTP login information
         """
-        def _log_progress(consumed: int, total: int):
-            # Only log every 20 seconds, to save log space...
-            nonlocal should_log
-            time_check = int(time.time())
-            if time_check % 20 == 0 and should_log:
-                log.info("Upload progress: " + "{:.2%}".format(consumed / total))
-                should_log = False
-            elif time_check % 20 != 0 and not should_log:
-                should_log = True
-
-        with self.sftp() as sftp:
-            # Ensure the folder path exists on C4
-            full_path = os.path.join(self._root_path(), folder_path)
-            traversed = []
-            for folder in full_path.split("/"):
-                if folder == "":
-                    continue
-
-                path_to_folder = os.path.join("/", "/".join(traversed), folder)
-                try:
-                    sftp.chdir(path_to_folder)
-                except IOError:
-                    sftp.mkdir(path_to_folder)
-                traversed.append(folder)
-
-            should_log = True
-            sftp.putfo(
-                file_obj,
-                os.path.join(full_path, file_name),
-                file_size=file_size,
-                callback=_log_progress)
\ No newline at end of file
+        # Ensure the folder path exists on C4
+        full_path = os.path.join(self._root_path(), folder_path)
+        traversed = []
+        for folder in full_path.split("/"):
+            if folder == "":
+                continue
+
+            path_to_folder = os.path.join("/", "/".join(traversed), folder)
+            try:
+                sftp.chdir(path_to_folder)
+            except IOError:
+                sftp.mkdir(path_to_folder)
+            traversed.append(folder)
+
+        cmd = f"lftp sftp://{cat_hook.connection.login}:{cat_hook.connection.password}@{cat_hook.connection.host}  -e \"get {source_file.full_path} -o {os.path.join(full_path, file_name)}; bye\""
```

Comment:
i'm, a little worried that this can allow for escaping issues.  A malicious user could inject arbitrary shell code?  into the login value here.  The fix would be to see if we could `exec_command` with an array of arguments rather than a singular string (in that case, no matter what, the interpolated parts are only considered part of one string argument to the exec, rather than potentially being invoked by the shell itself).

Reply:
Sure, I can make that change :+1:

Reply:
Hm, okay, so per the doc, [`exec_command` only take a string](https://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.exec_command).

Thoughts on using the same escaping utility as `subprocess`, (`list2cmdline`) a la https://stackoverflow.com/questions/3163236/escape-arguments-for-paramiko-sshclient-exec-command ?

Reply:
That makes sense.  Over ssh, you are relying on the configured shell of the remote user, so you don't know how the `exec` or `shell` processes will work, thus it is not possible to send the arguments as an array, only as a literal shell command.

Ok, can we atleast sanity check the interpolation parts against a regex?  Keeping it purely to alphanumeric?